packages/utils/telemetry-utils/src/logger.ts

```diff
@@ -483,30 +478,142 @@ export class PerformanceEvent {
- * - Helper class for error tracking that can be used to log an error in telemetry.
- * - Care needs to be taken not to log PII information!
- * - This allows additional properties to be logged because object of this instance will record all of their properties
- *   when logged with a logger.
- * - Logger ignores all properties from any other error objects (not being instance of LoggingError), with exception of
- *   'message' & 'stack' properties if they exists on error object.
- * - In other words, logger logs only what it knows about and has good confidence it does not container PII information.
+ * Type guard to identify if a particular value appears to be a tagged telemetry property
  */
-export class LoggingError extends Error {
+export function IsTaggedTelemetryPropertyValue(val: any): val is ITaggedTelemetryPropertyType {
+    const valAsTagged: ITaggedTelemetryPropertyType = val;
+    return (typeof(valAsTagged?.value) !== "object" && typeof(valAsTagged?.tag) === "number");
+}
+
+/**
+ * An error object that supports exporting its properties to be logged to telemetry
+ */
+export interface ILoggingError extends Error {
+    /** Return all properties from this object that should be logged to telemetry */
+    getTelemetryProperties(): ITelemetryProperties;
+}
+export const IsILoggingError = (x: any): x is ILoggingError => typeof x.getTelemetryProperties === "function";
+
+/**
+ * Helper class for error tracking that can be used to log an error in telemetry.
+ * The props passed in (and any set directly on the object after the fact) will be
+ * logged in accordance with the given TelemetryDataTag, if present.
+ *
+ * PLEASE take care to properly tag properties set on this object
+ */
+export class LoggingError extends Error implements ILoggingError {
     constructor(
         message: string,
-        props?: ITelemetryProperties,
+        props?: ITaggableTelemetryProperties,
     ) {
         super(message);
+        this.addTelemetryProperties(props);
+    }
+
+    /**
+     * Add additional properties to be logged
+     */
+    public addTelemetryProperties(props?: ITaggableTelemetryProperties) {
         Object.assign(this, props);
     }
 
-    // Return all properties
+    /**
+     * Enumerate this object's own properties to find those fit for telemetry.
+     * This protects us when someone sets a non-loggable field on this object (via 'as any')
+     */
+    private getOwnValidTelemetryProps(): ITaggableTelemetryProperties {
+        const props: ITaggableTelemetryProperties = {};
+        for (const key of Object.keys(this)) {
+            const val = this[key];
+            switch (typeof val) {
+                case "string":
+                case "number":
+                case "boolean":
+                case "undefined":
+                    props[key] = val;
+                    break;
+                default: {
+                    if (IsTaggedTelemetryPropertyValue(val)) {
+                        props[key] = val;
```

Comment:
i think we should include the tag in the key, either key_tag, or tag_key. also why i prefer a string for the tag